### ISSUE TITLE
startx display manager does not support autoLogin.

### ISSUE THREAD

**Issue by alpha-bit** [NixOS/nixpkgs#177924]
### Describe the bug

In options at "services.xserver.displayManager", when i "startx.enable=true;" and "autologin.enable=true;" ... nothing happens... :(

Previously tried to (wrongly) post this to nixos-search, 
so there is already some comments over this here:

  https://github.com/NixOS/nixos-search/issues/492

### Steps To Reproduce
1. as explained above

### Expected behavior

I expected the startx would be started, and restarted when ever exited the window manager.

Also would be nice to have a ...displayManager.startx.xinitrc option.

### Screenshots
no applicable.

### Additional context
no context.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

**Comment by pennae** [comment]
startx is not a display manager in the traditional sense, it only makes `startx` available for running an X session from a tty. you'll have to set autologin in your tty config instead and add `startx` to the appropriate shell startup file to get the desired effect.

**Comment by alpha-bit** [comment]
so it should not be classified as such, if i just want the startx command i would just install the package... right?...

and that would be ok if there is a way to login on tty1 only, where i can detect if i'm at tty and is so call startx.

O.o ... but this has been talked about, have you read https://github.com/NixOS/nixos-search/issues/492?...

if startx.enable is being classified as desktopManager, then should also support autologin and autorun. And run in a restarting loop, startx could be renamed to startx-loop with a startx-loop.xinitrc and a startx-loop.between=' a sh that runs between restarts and let's me less the x output'.

And if i wish just to have startx at console, i'll just add it to the list of packages.

So, how can i autologin at tty1 only?...
or how can i just run my-startx under my user after booting?...
or will startx get fixed as a pseudo desktopManager?...

i just want to autologin ::: it's sad and hurts to have lightdm installed just for that... :( ... i also don't need neither want lightdm asking me for user/pass when i exit the window manager...

Below is what i now use in the debian i'm migrating from.

# To be used at ~/.bash_login,
# typically like:
#
#   eval "`$ABS/bash/maybe_startx x`"

readonly ABS_started_x='$1' ;
export ABS_started_x ;

## maybe start X

printf "\nStart X... y ?  " ;

read -n1 -t \
  ${ABS_X_timeout_to_startx:-1} a ||
    a=y ;  # timed out
echo ;

[ -z "$a" -o "$a" = y ] &&
{
  xoutput=`startx 2>&1` ;

  ## X exited, maybe exit shell?

  printf "\nExit shell... y  ?  " ;

  read -n1 -t \
    ${ABS_X_timeout_to_shell:-3} a ||
      a=y ;  # timed out
  echo ;

  [ -z "$a" -o "$a" = y ] && exit ;

  ## Enter shell, maybe view startx output?

  printf "\nView startx output... y  ?  " ;

  read -n1 a ; echo ;

  [ -z "$a" -o "$a" = y ] &&

    printf %s "$xoutput" | less ;

} ;

trap - 2 20 ;  # Enable CTRL+C/Z

**Comment by alpha-bit** [comment]
O.o ... hoops # comment turned into bold... O.o?... lolz... i'm so noob at this.

**Comment by pennae** [comment]
autologin for just one tty isn't easy at the moment. you could conceivably [write a service](https://nixos.org/manual/nixos/stable/#sec-writing-modules) that prepares tty1 with the appropriate permissions and then runs your startx wrapper afterwards, but hooking that up correctly may require some experimentation

if you've just switched from another distro you might be best off using lightdm for now

**Comment by S-NA** [comment]
> how can i just run my-startx under my user after booting?

Are you dead set on running it on VT1? The first few VTs have a default getty running and would require some additional work to get X running with the method given below. If not you could use a modified version of the below systemd service.

```nix
  systemd.defaultUnit = "graphical.target";
  systemd.services.autostartx = {
    enable = true;
    description = "X11 session for <user>";
    after = [ "graphical.target" "systemd-user-sessions.service" ];
    wantedBy = [ "graphical.target" ];
    serviceConfig = {
      User = "<user you want to run as>";
      WorkingDirectory = "~";
      PAMName = "login";
      Environment = [ "XDG_SESSION_TYPE=x11" ];
      TTYPath = "/dev/tty8";
      StandardInput = "tty";
      UnsetEnvironment = "TERM";
      UtmpIdentifier = "tty8";
      UtmpMode = "user";
      StandardOutput = "journal";
      ExecStartPre = "${pkgs.kbd}/bin/chvt 8";
      ExecStart = "${pkgs.xorg.xinit}/bin/startx";
      Restart = "always";
      RestartSec = "3";
    };
  };
```

That being said not using a display manager is an esoteric thing to do and things tend to break see #94375 for example. You have to take care to ensure what is typically handled by the display-manger is done by your `.xinitrc`/setup.

**Comment by alpha-bit** [comment]
O.o ... "not using a display manager is an esoteric thing" ::: i have no need for one, is the right thing to not use one... :)

I did like that script you wrote... :)
... and it's now working... :) ... thank you... :)... here is my version:

`  systemd.defaultUnit = "graphical.target";
  systemd.services.autostartx = {
    enable = true;
    description = "X11 session for <user>";
    after = [ "graphical.target" "systemd-user-sessions.service" ];
    wantedBy = [ "graphical.target" ];
    serviceConfig = {
      # User = "<user you want to run as>";
      User = "r";
      WorkingDirectory = "~";
      PAMName = "login";
      Environment = [
        "XDG_SESSION_TYPE=x11"
        "TERM=xterm-256color"
          # avoid error from `less'
          # WARNING: terminal is not fully functional
      ];
      TTYPath = "/dev/tty8";
      StandardInput = "tty";

      # UnsetEnvironment = "TERM";

      UtmpIdentifier = "tty8";
      UtmpMode = "user";

      # StandardOutput = "journal";
      StandardOutput = "tty";
        # now i can see with less

      StandardError = "tty";
      ExecStartPre = "${pkgs.kbd}/bin/chvt 8";
      # ExecStart = "${pkgs.xorg.xinit}/bin/startx";
      ExecStart =
        "/home/r/.xmonad/abxm/lib/shell/bash/service-startx";
      Restart = "always";
      RestartSec = "3";
    };
  };`

# and service-startx script:

readonly ABS_started_x='$1' ;
export ABS_started_x ;

xoutput=`startx 2>&1` ;

## X exited, maybe exit shell?

printf "\nExit shell... y  ?  " ;

read -n1 -t \
  ${ABS_X_timeout_to_shell:-3} a ||
    a=y ;  # timed out
echo ;

# echo $a ; sleep 7;

[ -z "$a" -o "$a" = y ] && exit ;

## Enter shell, maybe view startx output?

printf "\nView startx output... y  ?  " ;

read -n1 a ; echo ;

[ -z "$a" -o "$a" = y ] &&

  printf %s "$xoutput" | less ;

**Comment by alpha-bit** [comment]
And i also disabled displayManager.startx, and lightdm :

    # startx.enable = true; # WHEN IT ACTUALLY WORKS!!!
    lightdm.enable = false;

While just adding xorg.xinit to the users..packages ... :)

& as i said before, if displayManager.startx just enables the presence of startx, then is redundant and i rather add it to the package list... :) ... All this displayManager/startx issue should be reviewed and made to make more sense (but i'm only very noob at this).

**Comment by rnhmjoj** [comment]
> so it should not be classified as such, if i just want the startx command i would just install the package... right?...

No, it's not just that: out-of-the-box startx/xinit will not work (properly) because the X.org configuration is not linked to /etc/X11, also `xorgserver` won't be called with the arguments specified in `services.xserver`.

> if startx.enable is being classified as desktopManager, then should also support autologin and autorun. 

Why? Not even all the "proper" display managers support autologin.

> So, how can i autologin at tty1 only?...

I think you're all overthinking it. It's pretty simple to setup:
```
  services.getty.autologinUser = "alpha-bit";
  environment.loginShellInit = ''
    if test "$(tty)" = "/dev/tty1" && ! pgrep -f xserver; then
      startx &
    fi
  '';
```

**Comment by alpha-bit** [comment]
You are right startx.enable must be true, or at least the keyboard will not work (: as i mentioned in the previous post :)

... while "should also support" ::: does not mean it must.

... and using services.getty.autologinUser = "alpha-bit";
leaves all other consoles also open ::: as has been also mentioned... twice. third now... :)

you should read the thread... :)

I'm quite happy with the service.autostartx approach suggested by S-NA.

Is even better than having it at tty1.

Is the (missing) way to have an startx + autologin + relooped ::: because is ridiculous to have a lighdm for not to use it, and bump into user logins... o.O?... for my personal single-user machine i use xtrlock-pam at xmonad startup to enter the password... o.O?... only logging of it when wanting to fully reload the xmonad session that i develop and recompile often... o.O?... no time waste with unneeded display managers... :)

### I still would rather have a aditional option for this, like:

    ...displayManager.startxdm = {
      enable = true;
        # while requiring
            # displayManager.autoLogin.enable = true;
            # displayManager.autoLogin.user = "user";

        tty = "7";
        xinitrc = "# would be nice, for this service only, different than the mentioned bellow"
     };

### ... and would also be nice the:

    ...displayManager.startx.xinitrc = "this could be implemented already";

**Comment by alpha-bit** [comment]
Noticed that in the manual https://nixos.org/manual/nixos/stable/index.html#sec-x11-auto-login it says explicitly:

  "Every display manager in NixOS supports auto-login."

O.o ... missing a ::: except displayManager.startx, which does not yet support auto-login... :( ... lolz.

**Comment by alkeryn** [comment]
@S-NA hey thanks for your systemd unit however running it didn't succeed on my machine, just nothing happens and on the chosen tty i just have a black screen
startx alone works though.
